File: C_EXAMPLE/example_basic_ResetSensor.cpp
/**
 * @file example_basic_ResetSensor.cpp
 * @brief Example demonstrating how to reset joint torque sensor offsets
 *        for the L1 robot arm.
 *
 * This example shows how to:
 * - Query SDK and controller versions
 * - Recover the robot arm from ERROR state
 * - Move the arm to the zero position in POSITION mode
 * - Read current joint torque sensor feedback
 * - Reset torque sensor offsets to zero
 *
 * Resetting torque sensor offsets is essential for accurate torque control
 * and collision detection.
 *
 * @warning Before running this example, ensure that **no tool is mounted**
 *          on Arm0’s end effector. Otherwise, torque control behavior may
 *          become incorrect.
 */

#include "L1Robot.h"

#ifdef _WIN32
#define SLEEP_MS(ms) Sleep(ms)
#else
#define SLEEP_MS(ms) usleep((ms) * 1000)
#endif

/**
 * @brief Read current joint torque sensor feedback for ARM0.
 *
 * This function retrieves real-time torque sensor data from the robot
 * feedback interface.
 *
 * @param[out] sensor_fbk Array of 7 torque sensor values (Nm)
 */
void ReadArm0SensorFbk(double sensor_fbk[7])
{
    const ROBOT_RT* rt_ptr = FX_L1_Fbk_GetRT();
    if (rt_ptr == NULL)
    {
        return;
    }

    for (int i = 0; i < 7; i++)
    {
        sensor_fbk[i] =
            rt_ptr->m_ARMS[0].m_ARM_OUT.m_ARM_FBK_Joint_SensorTor[i];
    }
}

/**
 * @brief Entry point of the torque sensor reset example.
 *
 * Workflow overview:
 * 1. Initialize communication with the robot controller
 * 2. Retrieve SDK and controller versions
 * 3. Check and recover the arm state (ERROR → IDLE)
 * 4. Move ARM0 to the zero position using POSITION mode
 * 5. Read current torque sensor feedback
 * 6. Reset torque sensor offsets to zero
 * 7. Read sensor feedback again to verify the result
 *
 * @param[in] argc Argument count (unused)
 * @param[in] argv Argument vector (unused)
 * @return int Exit code (0 on normal exit)
 */
int main(int argc, char** argv)
{
    int sdk_version = 0;                      ///< SDK version number
    int controller_version = 0;               ///< Controller firmware version
    FXStateType obj_state = FX_STATE_UNKNOWN; ///< Current arm state
    unsigned int system_errorcode = 0;        ///< Last system error code

    double zero_position[7] = { 0 };          ///< Zero joint position (degrees)
    double sensor_fbk[7] = { 0 };            ///< Torque sensor feedback values

    /* Get SDK version */
    sdk_version = FX_L1_System_GetSDKVersion();
    printf("SDK version is 0x%08x\n", sdk_version);

    printf("----------------------------------------- IMPORTANT ------------------------------------------\n");
    printf("Before running this example, please ensure that no tool is mounted on Arm0's end effector.\n");
    printf("Otherwise, Arm0 will not operate correctly in torque control mode!\n");
    printf("-------------------------------------------------------------------------------------------------\n");

    /* Establish communication with the controller */
    if (FX_L1_System_Link(6, 6, 7, 190, FX_LOG_ALL_FLAG) < 0)
    {
        printf("Failed to link system\n");
        goto WAIT_EXIT;
    }

    /* Get controller version */
    controller_version = FX_L1_System_GetControllerVersion();
    printf("Controller version is 0x%08x\n", controller_version);

    /* Check current arm state */
    obj_state = FX_L1_Fbk_CurrentState(FX_OBJ_ARM0);
    if (obj_state == FX_STATE_ERROR)
    {
        printf("Reset sensor offset must be run in FX_STATE_IDLE state, try to reset error\n");
        if (FX_L1_State_ResetError(FX_OBJ_ARM0, 1000, &system_errorcode) == FUNC_RET_SUCCESS)
        {
            printf("Reset arm0 error success, arm0 is now in STATE_IDLE state\n");
        }
        else
        {
            printf("Failed to reset arm0 error, errorcode = 0x%08x\n", system_errorcode);
            goto WAIT_EXIT;
        }
    }
    else if (obj_state != FX_STATE_IDLE)
    {
        if (FX_L1_State_SwitchToIdle(FX_OBJ_ARM0, 1000) != FUNC_RET_SUCCESS)
        {
            printf("Failed to transfer arm0 to STATE_IDLE state\n");
            goto WAIT_EXIT;
        }
    }

    /* Switch to POSITION mode */
    printf("Press any key to transfer arm0 to FX_STATE_POSITION state\n");
    getchar();
    if (FX_L1_State_SwitchToPositionMode(FX_OBJ_ARM0, 2000, 10, 10) != FUNC_RET_SUCCESS)
    {
        printf("Transfer arm0 to FX_STATE_POSITION state failed\n");
        goto WAIT_EXIT;
    }

    /* Move arm to zero position */
    printf("Press any key to let arm0 move to zero position\n");
    getchar();
    if (FX_L1_Runtime_SetJointPosCmd(1, FX_OBJ_ARM0, zero_position) != FUNC_RET_SUCCESS)
    {
        printf("Failed to set arm0's target position\n");
        goto WAIT_EXIT;
    }

    /* Return to IDLE state */
    printf("Please wait until arm0 reaches zero position, then press any key to transfer arm0 to FX_STATE_IDLE state\n");
    getchar();
    if (FX_L1_State_SwitchToIdle(FX_OBJ_ARM0, 1000) != FUNC_RET_SUCCESS)
    {
        printf("Transfer arm0 to FX_STATE_IDLE state failed\n");
    }

    /* Read torque sensor feedback before reset */
    ReadArm0SensorFbk(sensor_fbk);
    printf("Current arm0's sensor feedback: "
           "{%.4lf,%.4lf,%.4lf,%.4lf,%.4lf,%.4lf,%.4lf}\n",
           sensor_fbk[0], sensor_fbk[1], sensor_fbk[2],
           sensor_fbk[3], sensor_fbk[4], sensor_fbk[5],
           sensor_fbk[6]);

    /* Reset torque sensor offsets */
    printf("Press any key to clear sensor offset to zero\n");
    getchar();
    if (FX_L1_Config_ResetSensorOffset(FX_OBJ_ARM0) != FUNC_RET_SUCCESS)
    {
        printf("Clear arm0's sensor offset to zero failed\n");
        goto WAIT_EXIT;
    }

    SLEEP_MS(10);

    /* Read torque sensor feedback after reset */
    ReadArm0SensorFbk(sensor_fbk);
    printf("Clear sensor offset success, current arm0's sensor feedback: "
           "{%.4lf,%.4lf,%.4lf,%.4lf,%.4lf,%.4lf,%.4lf}\n",
           sensor_fbk[0], sensor_fbk[1], sensor_fbk[2],
           sensor_fbk[3], sensor_fbk[4], sensor_fbk[5],
           sensor_fbk[6]);

WAIT_EXIT:
    printf("Press any key to exit\n");
    getchar();
    return 0;
}

File: PYTHON_EXAMPLE/example_basic_ResetSensor.py
"""
@file example_basic_ResetSensor.py
@brief Example demonstrating how to reset joint torque sensor offsets
       for the L1 robot arm.

Workflow overview:
    1. Initialize communication with the robot controller
    2. Retrieve SDK and controller versions
    3. Run to zero joints in position state(Check and recover the arm state (ERROR → IDLE))
    4. Read current joint torque sensor feedback
    5. Reset torque sensor offsets to zero

Resetting torque sensor offsets is essential for accurate torque control
and collision detection.

@warning Before running this example, ensure that **no tool is mounted**
         on Arm0’s end effector. Otherwise, torque control behavior may
         become incorrect.

@warning Press Enter key at any time to stop the robot and exit.
"""

import sys
import time
import threading
from pathlib import Path

root_dir = Path(__file__).parent.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))
from PYTHON_SDK.GentoRobot import GentoRobot, FXLogMask, FXObjMask, FXObjType,state_map,error_dict

def main():
    def emergency_stop_thread(robot):
        input()
        print("\n[Emergency stop] Triggered. Stopping robot...")
        robot.emergency_stop(FXObjMask.OBJ_ALL_FLAG)
        time.sleep(0.1)
        print("Exiting program.")
        sys.exit(0)

    ctrl_obj=FXObjType.OBJ_ARM0
    target_state="Position"
    axis_mask = 0x7F
    arm_idx=0
    if ctrl_obj==FXObjType.OBJ_ARM1:
        arm_idx=1
    log_mask=FXLogMask.FX_LOG_INFO_FLAG
    vel=10
    acc=10
    pos1=[0]*7

    robot = GentoRobot()
    threading.Thread(target=emergency_stop_thread, args=(robot,), daemon=True).start()

    print(f"\n### 1/8 .Connecting...")
    ret = robot.link(6, 6, 7, 190, log_level=log_mask)
    if not robot._connected:
        print(f"Link failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 2/8. Getting versions...")
    print(f"Controller version:{robot.get_controller_version()}")
    print(f"Sdk version:{robot.get_sdk_version()}")

    print(f"\n### 3/8. Position state switching...")
    arm0_state = robot.current_state(ctrl_obj)
    if state_map[arm0_state] == "Error":
        ret, system_errorcode = robot.reset_error(ctrl_obj, 1000)
        if ret != 0:
            print(f"Reset error failed. Error msg: {error_dict[system_errorcode]}")
            return
    elif state_map[arm0_state] != "IDLE":
        ret = robot.switch_to_idle(ctrl_obj, 1000)
        if ret != 0:
            print(f"Switch to IDLE failed. Error msg: {robot._get_operate_error_msg(ret)}")
            return

    if state_map[arm0_state] != target_state:
        ret=robot.switch_to_position_mode(ctrl_obj,2000,vel,acc)
        if ret !=0:
            print(f"Error msg: {robot._get_operate_error_msg(ret)}")
            return

        rt_dict = robot.get_rt_dict()
        print(f"current state：{rt_dict["arms"][arm_idx]["state"]["cur"]}")
        print(f"current joints:{rt_dict["arms"][arm_idx]["fb"]["fb_pos"]}")
        sg_dict = robot.get_sg_dict()
        print(f"current vel:{sg_dict["arms"][arm_idx]['set']["vel_ratio"]}")
        print(f"current acc:{sg_dict["arms"][arm_idx]['set']["acc_ratio"]}")

    else:
        print(f"Already in {target_state} state, check vel and acc")
        sg_dict = robot.get_sg_dict()
        print(f"current vel:{sg_dict["arms"][arm_idx]['set']["vel_ratio"]}")
        print(f"current acc:{sg_dict["arms"][arm_idx]['set']["acc_ratio"]}")

        if int(sg_dict["arms"][arm_idx]['set']["vel_ratio"])!=vel:
            print(f"Set vel to {vel}")
            ret=robot.runtime_set_vel_ratio(ctrl_obj,vel)
            if ret!=0:
                print(f"set vel failed. Error msg: {robot._get_operate_error_msg(ret)}")
                return
            time.sleep(0.01)
            sg_dict = robot.get_sg_dict()
            print(f"After set, current vel:{sg_dict["arms"][0]['set']["vel_ratio"]}")

        if int(sg_dict["arms"][0]['set']["acc_ratio"])!=acc:
            print(f"Set acc to {acc}")
            ret=robot.runtime_set_acc_ratio(ctrl_obj,acc)
            if ret!=0:
                print(f"set acc failed. Error msg: {robot._get_operate_error_msg(ret)}")
                return
            time.sleep(0.01)
            sg_dict = robot.get_sg_dict()
            print(f"After set, current acc:{sg_dict["arms"][0]['set']["acc_ratio"]}")

    print(f"\n### 4/8. Run zero joints...")
    ret = robot.runtime_set_joint_pos_cmd(ctrl_obj, pos1)
    if ret != 0:
        print(f"Error msg: {robot._get_operate_error_msg(ret)}")
        return
    while 1:
        rt_dict = robot.get_rt_dict()
        fb_joints = rt_dict["arms"][arm_idx]["fb"]["fb_pos"]
        if robot.check_sequences_approx_equal(fb_joints, pos1):
            break
        time.sleep(0.001)
    print(f"Arm0 reached at {fb_joints}")
    time.sleep(0.5)

    print(f"\n### 5/8. Switching to IDLE...")
    ret = robot.switch_to_idle(ctrl_obj, 2000)
    if ret != 0:
        print(f"Error msg: {robot._get_operate_error_msg(ret)}")
        return
    arm0_state = robot.current_state(ctrl_obj)
    if state_map[arm0_state]!="IDLE":
        print(f"switch to idle faile")
        return
    print(f"Arm0 in idle state: {state_map[arm0_state]}")

    print(f"\n### 6/8. Reset sensor torque...")
    rt_dict = robot.get_rt_dict()
    fb_sensor = rt_dict["arms"][arm_idx]["fb"]["fb_sensor"]
    print(f"current sensor torque:{fb_sensor}")
    ret = robot.config_clear_sensor_offset(ctrl_obj)
    if ret != 0:
        print(f"Error msg: {robot._get_operate_error_msg(ret)}")
        return
    time.sleep(1)

    print(f"\n### 7/8. Read sensor torque...")
    rt_dict = robot.get_rt_dict()
    fb_sensor = rt_dict["arms"][arm_idx]["fb"]["fb_sensor"]
    print(f"After reset, current sensor torque:{fb_sensor}")

    print(f"\n### 8/8. Task finished.")

if __name__ == "__main__":
    main()
